Build OSX Catalina 10.15.1 (Xcode: 11.2 beta 2)

BlackDady opened the pull request:
Hi, i built a 2.14.3 version for OSX Catalina and i add a bash build_osx.sh

The content of it is base on the issue [#1931](https://github.com/hexchat/hexchat/issues/1931#issuecomment-390948491)

Special Thks to @philicious for it

Thks to share it and thks for your work for all the project.


My Build is there : [2.14.3 OSX Catalina](https://github.com/BlackDady/hexchat/releases/tag/v2.14.3)

<img width="250" alt="hexchat-Catalina" src="https://user-images.githubusercontent.com/57327197/72924563-73a41580-3d51-11ea-87d5-a3309035f1cf.png">

TingPing commented:
While I'm glad it builds for you I'm not a fan of keeping a big build script in tree that honestly won't get maintained. Also using brew isn't acceptable for releasing binaries like you have since they aren't portable to multiple macos versions.

BlackDady commented:
Hi, OK i modified the script to make a portable version also.
It's the first time that i do it but i removed my libssl.dylib of my build folder ($Home/lib) and the portable version seems working.

hexchat binary libraries:
```
/build/hexchat-portable.app/Contents/MacOS$ otool -L hexchat
hexchat:
@executable_path/../Resources/lib/libgio-2.0.0.dylib (compatibility version 6201.0.0, current version 6201.4.0)
@executable_path/../Resources/lib/libgobject-2.0.0.dylib (compatibility version 6201.0.0, current version 6201.4.0)
@executable_path/../Resources/lib/libglib-2.0.0.dylib (compatibility version 6201.0.0, current version 6201.4.0)
@executable_path/../Resources/lib/libintl.8.dylib (compatibility version 10.0.0, current version 10.6.0)
@executable_path/../Resources/lib/libgmodule-2.0.0.dylib (compatibility version 6201.0.0, current version 6201.4.0)
@executable_path/../Resources/lib/libgtk-quartz-2.0.0.dylib (compatibility version 2401.0.0, current version 2401.32.0)
@executable_path/../Resources/lib/libgdk-quartz-2.0.0.dylib (compatibility version 2401.0.0, current version 2401.32.0)
@executable_path/../Resources/lib/libpangocairo-1.0.0.dylib (compatibility version 4401.0.0, current version 4401.7.0)
@executable_path/../Resources/lib/libpango-1.0.0.dylib (compatibility version 4401.0.0, current version 4401.7.0)
@executable_path/../Resources/lib/libgdk_pixbuf-2.0.0.dylib (compatibility version 4001.0.0, current version 4001.0.0)
@executable_path/../Resources/lib/libnotify.4.dylib (compatibility version 5.0.0, current version 5.0.0)
@executable_path/../Resources/lib/libssl.1.1.dylib (compatibility version 1.1.0, current version 1.1.0)
@executable_path/../Resources/lib/libcrypto.1.1.dylib (compatibility version 1.1.0, current version 1.1.0)
@executable_path/../Resources/lib/libproxy.1.dylib (compatibility version 1.0.0, current version 1.0.0)
@executable_path/../Resources/lib/libcanberra.0.dylib (compatibility version 3.0.0, current version 3.5.0)
@executable_path/../Resources/lib/libdbus-glib-1.2.dylib (compatibility version 6.0.0, current version 6.4.0)
/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.0.0)

```


The script is not a big script (to see Main()) : 
The first part of the script install the environment  to build (install librairies from brew only if is needed and environment variables (LD_PATH...) only if is need (like brew version of gettext or openssl that not linked with /usr/local/lib by default)

The second part call meson and ninja

The Thrid part make the bundles of hexchat.app and hexchat-portable.app

There are a few lines codes for patch some details.

It was the first time that i use meson and i spent a lot of days to understand how build, read documentation , understand that jhbuild is a old process for build it.

I taked the choice for don't touch the exiting code or meson.build files because i am not an expert and i don't want to maintain after ;)

But if you want i can explain to you which modifications is needed into the meson.build file for remove the patch lines of the script.

So if you can add this Script V1 into your tree could be nice. For me for the future release, and maybe for another people but that script and the portable bundle needs to be tested by someone else.

Also the part in the OSX folder of your tree (to make the OSX bundle) is Out of date if i understand because that continue to use jhbuild Right ?
 
Big Thks for your work.

philicious commented:
great stuff @BlackDady ! I never got around finishing the portable version or doing a proper build script ..

@TingPing I think adding this build script to the repo is really worth it and many people will be glad about it in the future when they wanna build a MacOS version. Even if it would become stale, there will certainly be someone picking it up again

TingPing commented:
Move it into the `osx` dir, print a warning at the start of the script that it is unsupported and may not work, and I'll accept it as-is.

BlackDady commented:
OK no problem i can move it under the OSX folder. For the moment i have 2 works to do to make a really portable version :
- First is understand (because for the moment i don't understand lol) how to put the share folder in relative of the main binairy path for example .. (e.g. support of multi-language .po)
- Second is put the model of the script Launcher of "GIMP" to make a very full portable version (include, python environment variables and binaries, start service like dbus .....)

I have patch the trouble of lua big usage memory. The problem was it : In OSX 64bit the default pagezero_size is 64GB, and you have to fixit on the link command of the main binary only (option don't support for a .dylib but just the final binary (with main -Wl,-pagezero_size,0x10000) ). But maybe someone could be fixit on the src meson.build ?


So when the dbus service is running, you can see the icon on the title bar but if you open the window, when the widows is closed the application crash: 
```
/applications/HExChat.app/Contents/MacOS$ hexchat

(hexchat:29295): libnotify-WARNING **: 12:27:27.034: Failed to connect to proxy
Segmentation fault: 11


```

![Capture d’écran 2020-01-28 à 12 32 08](https://user-images.githubusercontent.com/57327197/73260458-3c5abc00-41ca-11ea-9479-35f2cdbac148.png)

If the trouble is not again fixed on the master branch, if you explain me how to active supply of log messages report to you.

And a last thing, because i check how to fix-it but i think GTK2.0 don't support Hi resolution Screen like Macbook Retina, so .... a new version with GTK3.0 in your basket ?

Thks

TingPing commented on osx/certs/README.TXT:
```diff
@@ -0,0 +1,2 @@
+Add here Trust Root Certificate
+Warning: Don't put fucking false CA to hack me !
```
```suggestion
```

BlackDady replied:
Maybe, a day, i could put my own CA servers into this directory, but i could add a .gitignore for .pem or .crt ? It's for a standalone OpenSSL key chain ;)

BlackDady replied:
The problem is OSX don't use openSSL but DarwinSSL and i wanted a full standalone OpenSSL Version

TingPing commented on osx/certs/README.TXT:
```diff
@@ -0,0 +1,2 @@
+Add here Trust Root Certificate
```
Why is this even needed?

TingPing commented:
I feel like this branch is growing far beyond what I'm comfortable with merging..

BlackDady commented:
> I feel like this branch is growing far beyond what I'm comfortable with merging..

For my modifies, i have added new files mostly. Except  a modify for the StatusIcon Plugin.

But i have again one think/work to integreted in the MAC version, it's a TOR standlone proxy and maybe a new plugin to see and change the identity

BlackDady commented:
OK for me it's good now with this version.
i still have to make a plugin to pilot Tor but i don't have the time now.
I need somebody to test it on OSX without brew packages and without xcode.
Maybe could i open an issue on your repository to ask to test to my last bundle ? am i authorize to do it ?

philicious commented:
@BlackDady I'm happy to help with testing. I happen to have a brand-new, nothing-installed Macbook Pro at my hands. Its at 10.14 Mojave which I can update to Catalina. 
where can I find your latest build ?

BlackDady commented:
@philicious Hi,
OK Nice news.


You can Find my Build there: [2.14.3 OSX Catalina](https://github.com/BlackDady/hexchat/releases/tag/v2.14.3)

To Help me, after download my build, launch the application with a terminal session like :
`/Applications/hexchat.app/Contents/MacOS$ hexchat`
And give me all the printed lines

A Friend have tried to run it on Mojave and for the moment there are troubles.
But i can try, with you, to resolv it if you want.

The problem is this:
```
dyld: lazy symbol binding failed: Symbol not found: _objc_opt_class
  Referenced from: /Applications/hexchat.app/Contents/Resources/lib/libgdk-quartz-2.0.0.dylib (which was built for Mac OS X 10.15)
  Expected in: /usr/lib/libobjc.A.dylib
```


The Librairies of /usr/lib is not include in the bundle like libobjc.A.dylib
BUT if you have GIMP.app you can try it for me:

```
do it under : /Applications/hexchat.app/Contents/
ln -s ./Resources/lib .
and put the  ......./GIMP.app/Contents/Resources/lib/libgdk-quartz-2.0.0.dylib  into your
/Applications/hexchat.app/Contents/lib/
```
If you have again a problem with another GTK+ lib, do the same thing.

And Tell me more ;)

PS: Replace /Applications by the path where is the bundle ;)

philicious commented:
@BlackDady isnt that Release you link to 26days old but you changed/updated things after that? at least it seems so looking at this PR

BlackDady commented:
@philicious 
Yes the hexchat.app.zip was make there were 5 days ago.
This Bundle include a standalone TOR proxy on The port 9250, that you can use in hexchat like Modatory sever (proxy sock 5) on 127.0.0.1:9250

BlackDady commented:
@philicious
Hi, i add a hexchat-mojave.app.zip
This version should be compatible with mojave
can you try ?

pasamio commented:
I tried the hexchat mojave app and it launched and connects to an IRC server. I'm not sure I agree with autostarting Tor out of nowhere though.

BlackDady commented:
> I tried the hexchat mojave app and it launched and connects to an IRC server. I'm not sure I agree with autostarting Tor out of nowhere though.

Hi, Thks for your feedback. 
Tor is started but not use if you didn't configure a mandatory proxy server to use it into hexchat.
However if you want don't start Tor, you can modify the startup script: hexchat.app/Contents/MacOS/hexchat, at the end, just add an # infront of the line "$bundle_bin/tor" --defaults-torrc ...
